File: mifosng-android/src/main/java/com/mifos/mifosxdroid/views/SignatureView.java
package com.mifos.mifosxdroid.views;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.os.Environment;
import androidx.annotation.Nullable;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.MotionEvent;
import android.view.View;

import com.mifos.mifosxdroid.R;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.util.ArrayList;

/**
 * Created by Tarun on 28-06-2017.
 */

public class SignatureView extends View {

    private final String LOG_TAG;
    private Paint signPaint;
    private ArrayList<Float> mXCoordinateList;
    private ArrayList<Float> mYCoordinateList;
    private OnSignatureSaveListener mOnSignatureSaveListener;
    private Context mContext;

    public SignatureView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        LOG_TAG = getClass().getSimpleName();

        mContext = context;
        setDrawingCacheEnabled(true);
        Resources r = getResources();
        int strokeSize = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 5,
                r.getDisplayMetrics());

        signPaint = new Paint();
        signPaint.setColor(Color.BLACK);
        signPaint.setStrokeWidth(strokeSize);
        signPaint.setAntiAlias(true);
        mXCoordinateList = new ArrayList<>();
        mYCoordinateList = new ArrayList<>();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        canvas.drawColor(Color.WHITE);
        if (mXCoordinateList.size() > 2) {
            for (int i = 1; i < mXCoordinateList.size(); i++) {
                if (!(mXCoordinateList.get(i - 1) == -1.0 || mXCoordinateList.get(i) == -1)) {
                    canvas.drawLine(mXCoordinateList.get(i - 1), mYCoordinateList.get(i - 1),
                            mXCoordinateList.get(i), mYCoordinateList.get(i), signPaint);
                }
            }
        }
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        super.onTouchEvent(event);
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN: {
                mXCoordinateList.add(event.getX());
                mYCoordinateList.add(event.getY());
                break;
            }
            case MotionEvent.ACTION_UP: {
                mXCoordinateList.add((float) -1.0);
                mYCoordinateList.add((float) -1.0);
                break;
            }
            case MotionEvent.ACTION_MOVE: {
                mXCoordinateList.add(event.getX());
                mYCoordinateList.add(event.getY());
                invalidate();
                break;
            }
        }
        return true;
    }

    public interface OnSignatureSaveListener {
        void onSignSavedError(String errorMsg);

        void onSignSavedSuccess(String absoluteFilePath);
    }

    public void setOnSignatureSaveListener(OnSignatureSaveListener listener) {
        mOnSignatureSaveListener = listener;
    }

    public void saveSignature(int id) {
        String path = Environment.getExternalStorageDirectory() +
                mContext.getResources().getString(R.string.signature_image_directory);
        File signatureStorageDirectory = new File(path);
        File signatureFile;
        if (!signatureStorageDirectory.exists()) {
            boolean makeRequiredDirectories = signatureStorageDirectory.mkdirs();
            if (!makeRequiredDirectories) {
                mOnSignatureSaveListener.onSignSavedError(mContext.getResources().
                        getString(R.string.sign_dir_not_created_msg));
                return;
            }
        }

        try {
            Bitmap signatureBitmap = getDrawingCache();
            signatureFile = new File(signatureStorageDirectory.getPath(), id + ".jpeg");
            signatureBitmap.compress(Bitmap.CompressFormat.JPEG, 100,
                    new FileOutputStream(signatureFile));
        } catch (FileNotFoundException e) {
            if (mOnSignatureSaveListener != null) {
                mOnSignatureSaveListener.onSignSavedError(e.getMessage());
            }
            return;
        }
        if (mOnSignatureSaveListener != null) {
            mOnSignatureSaveListener.onSignSavedSuccess(signatureFile.getAbsolutePath());
        }
    }


    public void clear() {
        mXCoordinateList.clear();
        mYCoordinateList.clear();
        destroyDrawingCache();
        invalidate();
    }

    public int getXCoordinateSize() {
        return mXCoordinateList.size();
    }

    public int getYCoordinateSize() {
        return mYCoordinateList.size();
    }
}

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/views/SignatureView.kt
package com.mifos.mifosxdroid.views

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.os.Environment
import android.util.AttributeSet
import android.util.TypedValue
import android.view.MotionEvent
import android.view.View
import com.mifos.mifosxdroid.R
import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream

/**
 * Created by Tarun on 28-06-2017.
 */
class SignatureView(context: Context, attrs: AttributeSet?) : View(context, attrs) {
    private val LOG_TAG: String = javaClass.simpleName
    private val signPaint: Paint
    private val mXCoordinateList: ArrayList<Float>
    private val mYCoordinateList: ArrayList<Float>
    private var mOnSignatureSaveListener: OnSignatureSaveListener? = null
    private val mContext: Context

    init {
        mContext = context
        isDrawingCacheEnabled = true
        val r = resources
        val strokeSize = TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP, 5f,
            r.displayMetrics
        ).toInt()
        signPaint = Paint()
        signPaint.color = Color.BLACK
        signPaint.strokeWidth = strokeSize.toFloat()
        signPaint.isAntiAlias = true
        mXCoordinateList = ArrayList()
        mYCoordinateList = ArrayList()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        canvas.drawColor(Color.WHITE)
        if (mXCoordinateList.size > 2) {
            for (i in 1 until mXCoordinateList.size) {
                if (!(mXCoordinateList[i - 1].toDouble() == -1.0 || mXCoordinateList[i] == -1f)) {
                    canvas.drawLine(
                        mXCoordinateList[i - 1], mYCoordinateList[i - 1],
                        mXCoordinateList[i], mYCoordinateList[i], signPaint
                    )
                }
            }
        }
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        super.onTouchEvent(event)
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                mXCoordinateList.add(event.x)
                mYCoordinateList.add(event.y)
            }

            MotionEvent.ACTION_UP -> {
                mXCoordinateList.add(-1.0.toFloat())
                mYCoordinateList.add(-1.0.toFloat())
            }

            MotionEvent.ACTION_MOVE -> {
                mXCoordinateList.add(event.x)
                mYCoordinateList.add(event.y)
                invalidate()
            }
        }
        return true
    }

    interface OnSignatureSaveListener {
        fun onSignSavedError(errorMsg: String)
        fun onSignSavedSuccess(absoluteFilePath: String)
    }

    fun setOnSignatureSaveListener(listener: OnSignatureSaveListener?) {
        mOnSignatureSaveListener = listener
    }

    fun saveSignature(id: Int) {
        val path = Environment.getExternalStorageDirectory().toString() +
                mContext.resources.getString(R.string.signature_image_directory)
        val signatureStorageDirectory = File(path)
        val signatureFile: File
        if (!signatureStorageDirectory.exists()) {
            val makeRequiredDirectories = signatureStorageDirectory.mkdirs()
            if (!makeRequiredDirectories) {
                mOnSignatureSaveListener?.onSignSavedError(mContext.resources.getString(R.string.sign_dir_not_created_msg))
                return
            }
        }
        try {
            val signatureBitmap = drawingCache
            signatureFile = File(signatureStorageDirectory.path, "$id.jpeg")
            signatureBitmap.compress(
                Bitmap.CompressFormat.JPEG, 100,
                FileOutputStream(signatureFile)
            )
        } catch (e: FileNotFoundException) {
            if (mOnSignatureSaveListener != null) {
                e.message?.let { mOnSignatureSaveListener?.onSignSavedError(it) }
            }
            return
        }
        if (mOnSignatureSaveListener != null) {
            mOnSignatureSaveListener?.onSignSavedSuccess(signatureFile.absolutePath)
        }
    }

    fun clear() {
        mXCoordinateList.clear()
        mYCoordinateList.clear()
        destroyDrawingCache()
        invalidate()
    }

    val xCoordinateSize: Int
        get() = mXCoordinateList.size
    val yCoordinateSize: Int
        get() = mYCoordinateList.size
}